src/Components/Server/src/Circuits/CircuitIdFactory.cs

```diff
@@ -0,0 +1,65 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Security.Cryptography;
+using Microsoft.AspNetCore.DataProtection;
+using Microsoft.AspNetCore.WebUtilities;
+using Microsoft.Extensions.Options;
+
+namespace Microsoft.AspNetCore.Components.Server.Circuits
+{
+    // This is a singleton instance
+    // Generates strong criptographic ids for circuits that are protected with authenticated encryption.
+    internal class CircuitIdFactory
+    {
+        // Follow up with blowdart/levi. Create should be good enough, but not sure if we need
+        // to do anything additional for FIPS compliance.
+        private readonly RandomNumberGenerator _generator = RandomNumberGenerator.Create();
+        private readonly IDataProtector _protector;
+
+        public CircuitIdFactory(IOptions<CircuitOptions> options)
+        {
+            _protector = options.Value.CircuitIdProtector;
+        }
+
+        // Generates a circuit id that is produced from a strong cryptographic random number generator
+        // we don't care about the underlying payload, other than its uniqueness and the fact that we
+        // authenticate encrypt it using data protection.
+        // For validation, the fact that we can unprotect the payload is guarantee enough.
+        public string CreateCircuitId()
```

Comment:
In the future we can pass additional information that we can validate, but this should be good enough as 32 bytes of entropy ensure us uniqueness